store/tikv/2pc.go

```diff
@@ -1050,3 +1008,83 @@ func appendBatchBySize(b []batchKeys, region RegionVerID, keys [][]byte, sizeFn
 	}
 	return b
 }
+
+// newTaskProcessor create processor to handle concurrent batch works(prewrite/commit etc)
+func newTaskProcessor(rateLimit int) *taskProcessor {
+	return &taskProcessor{newRateLimit(rateLimit)}
+}
+
+// startWork concurrently do the work for each batch considering rate limit
+func (tp *taskProcessor) startWorker(exitCh chan struct{}, ch chan error,
+	procFn procFunc, action twoPhaseCommitAction,
+	backoffer *Backoffer, batches []batchKeys) {
+	for idx, batch1 := range batches {
+		if exit := tp.rateLimiter.getToken(exitCh); !exit {
+			batch := batch1
+			go func(i int) {
+				var procRes error
+				defer tp.rateLimiter.putToken()
+				if action == actionCommit {
+					// Because the secondary batches of the commit actions are implemented to be
+					// committed asynchronously in background goroutines, we should not
+					// fork a child context and call cancel() while the foreground goroutine exits.
+					// Otherwise the background goroutines will be canceled execeptionally.
+					// Here we makes a new clone of the original backoffer for this goroutine
+					// exclusively to avoid the data race when using the same backoffer
+					// in concurrent goroutines.
+					singleBatchBackoffer := backoffer.Clone()
+					procRes = procFn(singleBatchBackoffer, batch)
+				} else {
+					singleBatchBackoffer, singleBatchCancel := backoffer.Fork()
+					defer singleBatchCancel()
+					procRes = procFn(singleBatchBackoffer, batch)
+				}
+				ch <- procRes
+			}(idx)
+		} else {
+			logutil.Logger(backoffer.ctx).Info("break startWorker", zap.Stringer("action", action))
+			break
+		}
+	}
+}
+
+// process will start worker routine and collect results
+func (tp *taskProcessor) process(committer *twoPhaseCommitter, procFn procFunc,
+	bo *Backoffer, action twoPhaseCommitAction, batches []batchKeys) error {
+	var err error
+	c := committer
+	// For prewrite, stop sending other requests after receiving first error.
+	backoffer := bo
```

Comment:
ditto.

Reply:
more clear to use two since there maybe fork operation